### PR TITLE
Fix #342 : multiple stores on the same database

### LINKED ISSUE #342: localforage.createInstance: cannot create several instances with the same name (=db name)
```js
var table1 = localforage.createInstance({name: 'db', storeName: 'store1' });
var table2 = localforage.createInstance({name: 'db', storeName: 'store2' });
//Do something with table1 and table2
```
This fails on Windows 8.1 + Chrome 40 (with latest updates as of 04/02/2015).
The same with different db names ('db1' and 'db2') works but you then have one IndexedDB database per table.

image

### PR DESCRIPTION
**Pull request by sebweaver**
Implementation summary for IndexedDB:
- Keep track on all running forages per database.
- Wait for them to be ready before processing any new connection to the same
database.
- Probe the current database to check if an upgrade is required (greater version
or new store).
- If so, upgrade the database by reopening it.
- Share the final connection (original or upgraded) amongts all running forages.

Implementation summary for localStorage:
- Prefix the key with the store name (after database), unless it's the default
name.

### CODE REVIEW

**Comment by thgreasi** [comment]
:+1:

File: src/drivers/indexeddb.js [hunk 1]
```diff
@@ -181,27 +181,164 @@
             }
         }
 
+        // Initialize a singleton container for all running localForages.
+        if (!globalObject.container) {
+            globalObject.container = {};
```
Comment (thgreasi) [inline src/drivers/indexeddb.js]: Is there a reason that we attach the `container` on the global object?? It feels too intrusive to me to just add an extra variable to the window object.

This could probably be just a local variable [around line 18](https://github.com/sebweaver/localForage/blob/issue-342-multiple-stores-on-same-db/src/drivers/indexeddb.js#L18), inside the closure of the IndexedDB driver.

As an alternative, it could be attached on the LocalForage class as a static property, named like `_dbContexts`. On the other hand I don't see a reason that this couldn't just be completely private to the IndexedDB driver.

File: src/drivers/localstorage.js [hunk 2]
```diff
@@ -39,6 +39,10 @@
 
         dbInfo.keyPrefix = dbInfo.name + '/';
 
+        if (dbInfo.storeName !== self._defaultConfig.storeName) {
+            dbInfo.keyPrefix += dbInfo.storeName + '/';
+        }
```
Comment (thgreasi) [inline src/drivers/localstorage.js]: Simply said, :+1: on this. I can't remember which issue it closes!?

File: test/test.config.js [hunk 3]
```diff
@@ -197,7 +197,7 @@ describe('Config API', function() {
                     });
                 });
             } else if (localforage.driver() === localforage.LOCALSTORAGE) {
-                var dbValue = JSON.parse(localStorage['My Cool App/some key']);
+                var dbValue = JSON.parse(localStorage['My Cool App/storeFront/some key']);
```
Comment (thgreasi) [inline test/test.config.js]: I would prefer not to use the 'storeFront' string in here, since it is related to several trademarked products.

File: src/drivers/indexeddb.js (hunk 1 above)
Comment (sebweaver) [inline src/drivers/indexeddb.js]: My bad, you're totally right about `globalObject`! It's the first thing I did on this PR and I forgot to change it. I will attach this container locally in the IndexedDB driver, name it `dbContexts`, as you suggest, and update the PR.

File: src/drivers/localstorage.js (hunk 2 above)
Comment (sebweaver) [inline src/drivers/localstorage.js]: No issue in particular AFAIK, but necessary to fulfill the new test case of this PR.

In fact, it's surprising the support of `storeName` hasn't been implemented before. At least in the same time than the two other drivers. In any case, I didn't see anything about that in the localStorage driver.

File: test/test.config.js (hunk 3 above)
Comment (sebweaver) [inline test/test.config.js]: Not my choice of name. :-)

I've just adapted the test to make it pass with the implementation of the `storeName` namespace of this PR. The original name was here:

https://github.com/sebweaver/localForage/blob/issue-342-multiple-stores-on-same-db/test/test.config.js#L78

Comment (thgreasi) [inline test/test.config.js]: Ok then!

File: src/drivers/localstorage.js (hunk 2 above)
Comment (thgreasi) [inline src/drivers/localstorage.js]: I remember someone had reported this somewhere....
Anyway, did you also add a test case for this?

Comment (sebweaver) [inline src/drivers/localstorage.js]: Test cases already exist here:

https://github.com/sebweaver/localForage/blob/issue-342-multiple-stores-on-same-db/test/test.api.js#L772

And there:

https://github.com/sebweaver/localForage/blob/issue-342-multiple-stores-on-same-db/test/test.api.js#L799

I just enriched them and they failed for localStorage too (not only IndexedDB). That's why I added the above implementation. Far simpler than for IndexedDB by the way... :-)

Comment (thgreasi) [inline src/drivers/localstorage.js]: Great :+1:

File: src/drivers/indexeddb.js [hunk 4]
```diff
@@ -181,27 +182,164 @@
...
+            self._dbInfo = dbInfo;
+            // Share the final connection amongst related localForages.
+            for (var k in forages) {
+                var forage = forages[k];
+                if (forage !== self) { // Self is already up-to-date.
+                    forage._dbInfo.db = dbInfo.db;
+                    forage._dbInfo.version = dbInfo.db;
+                }
+            }
+        });
```
Comment (thgreasi) [inline src/drivers/indexeddb.js]: Should we also add a `catch()` here and `reject` the promise returned by `_initStorage`?

Comment (sebweaver) [inline src/drivers/indexeddb.js]: No, there's no nested promise here, but a chain of existing promises which have been already resolved of rejected individually. If a rejection occurs in a chain of promises, it bubbles up and the whole chain is rejected.

Comment (thgreasi) [inline src/drivers/indexeddb.js]: Hmmmm
Lets say I create instance1 giving a very large amount of DB size and as a result the ready() promise of instance1 rejects at some point.
Now lets say that my code handles that error and creates instance2 with the same db parameters as instance1 but with a smaller & valid size for the db size option. 
Will that work? I think not, since we wait until all `ready()` promises of the related instances are **resolved** (even though your comment says "Initialize the connection process only when all the related localForages are ready").
I think that we just need to know that their promises are no longer pending and as a result their `_iniStorage` call has completely finished.

TL;DR: I think that instead of `ready.push(forage.ready());` in [Line 213](https://github.com/mozilla/localForage/pull/452/files#diff-29be772a36519980595b932d60929feeR213) we should use `ready.push(forage.ready().catch())` (if I'm not mistaken or understood anything wrong);

File: test/test.config.js (hunk 3 above)
Comment (tofumatt) [inline test/test.config.js]: We can certainly change it. I wrote that so long ago I can't remember why, and it's worth changing ;-)

I'll do that outside this PR (or when we merge it in): #453 is filed for it :smile:

**Comment by tofumatt** [comment]
Wow, this is _awesome_! :+1:

File: src/drivers/indexeddb.js [hunk 5]
```diff
@@ -181,27 +182,164 @@
...
+            // Register the new context in the global container.
+            dbContexts[dbInfo.name] = dbContext;
+        }
+
+        // Register itself as a running localForage in the current context.
+        dbContext.forages.push(this);
+
+        // Create an array of readiness of the related localForages.
+        var ready = [];
+        for (var j = 0; j < dbContext.forages.length; j++) {
```
Comment (tofumatt) [inline src/drivers/indexeddb.js]: Any reason you're using `j` over `i`? ^_^

File: src/drivers/indexeddb.js (hunk 4 above)
Comment (tofumatt) [inline src/drivers/indexeddb.js]: If there's no harm to @thgreasi's suggestion I _think_ I agree with him, but I'll admit my brain gets scrambled by Promises sometimes at this level.

If it would be possible to write a test to cover @thgreasi's use-case that would be great. Would it be as simple as:

```js
var databaseInfo = {
    name: 'localForaging',
    storeName: 'helloApp',
    size: 24903680
};

var db = localforage.createInstance(databaseInfo)

db.then(function() {
    // This should fail?
}, function() {
    databaseInfo.size = 5000;
    db = localforage.createInstance(databaseInfo);
});

Kind of a contrived example, but it should do the trick? Unless I misunderstand! ^_^

**Comment by tofumatt** [comment]
Other than the question of using `catch()` as @thgreasi mentioned here: https://github.com/mozilla/localForage/pull/452#discussion_r40494119

this is awesome and gets a :+1: r+ from me.

Once we figure out what needs to me done there we can merge it. I have a few nitpicks about variable names and stuff but we can fix those after merge; this PR is great :smile: 

Thanks @sebweaver!

File: src/drivers/indexeddb.js (hunk 4 above)
Comment (thgreasi) [inline src/drivers/indexeddb.js]: Hmmm.... (again),
Changing `ready.push(forage.ready());` in [line 213](https://github.com/mozilla/localForage/pull/452/files#diff-29be772a36519980595b932d60929feeR213) into `ready.push(forage.ready().catch());` should fix the aforementioned test case...
Also we should probably add a comment there explaining that we only care about the promise not being pending.
Too much effort for such a small change :smile:

Comment (tofumatt) [inline src/drivers/indexeddb.js]: Should we just add in the `catch` and then merge this into master and address this use case later if needed then?

I'd be okay with that... this patch is awesome.

Comment (thgreasi) [inline src/drivers/indexeddb.js]: Now you made me hesitant about the completeness and correctness of my
suggested fix...
This is a way too small test case to leave for an other PR. On the other
hand I feel ok with the current state of this PR.
How about giving it an extra day or two?

On Sun, Sep 27, 2015, 22:16 Matthew Riley MacPherson <
notifications@github.com> wrote:

> In src/drivers/indexeddb.js
> <https://github.com/mozilla/localForage/pull/452#discussion_r40508172>:
>
> > +                // Reopen the database for upgrading.
> > +                return _getUpgradedConnection(dbInfo);
> > +            }
> > +            return db;
> > +        }).then(function(db) {
> > +            dbInfo.db = dbContext.db = db;
> > +            self._dbInfo = dbInfo;
> > +            // Share the final connection amongst related localForages.
> > +            for (var k in forages) {
> > +                var forage = forages[k];
> > +                if (forage !== self) { // Self is already up-to-date.
> > +                    forage._dbInfo.db = dbInfo.db;
> > +                    forage._dbInfo.version = dbInfo.db;
> > +                }
> > +            }
> > +        });
>
> Should we just add in the catch and then merge this into master and
> address this use case later if needed then?
>
> I'd be okay with that... this patch is awesome.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/localForage/pull/452/files#r40508172>.
>
-- 
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

Comment (tofumatt) [inline src/drivers/indexeddb.js]: Cool, I can be patient. I’ll wait for a bit and check back in when it’s figured out :-)

- tofumatt

On 27 September 2015 at 20:26:39, Thodoris Greasidis (notifications@github.com) wrote:

In src/drivers/indexeddb.js:

> +                // Reopen the database for upgrading.
> +                return _getUpgradedConnection(dbInfo);
> +            }
> +            return db;
> +        }).then(function(db) {
> +            dbInfo.db = dbContext.db = db;
> +            self._dbInfo = dbInfo;
> +            // Share the final connection amongst related localForages.
> +            for (var k in forages) {
> +                var forage = forages[k];
> +                if (forage !== self) { // Self is already up-to-date.
> +                    forage._dbInfo.db = dbInfo.db;
> +                    forage._dbInfo.version = dbInfo.db;
> +                }
> +            }
> +        });
Now you made me hesitant about the completeness and correctness of my suggested fix... This is a way too small test case to leave for an other PR. On the other hand I feel ok with the current state of this PR. How about giving it an extra day or two?
On Sun, Sep 27, 2015, 22:16 Matthew Riley MacPherson < notifications@github.com> wrote: In src/drivers/indexeddb.js <https://github.com/mozilla/localForage/pull/452#discussion_r40508172>: > + // Reopen the database for upgrading. > + return _getUpgradedConnection(dbInfo); > + } > + return db; > + }).then(function(db) { > + dbInfo.db = dbContext.db = db; > + self._dbInfo = dbInfo; > + // Share the final connection amongst related localForages. > + for (var k in forages) { > + var forage = forages[k]; > + if (forage !== self) { // Self is already up-to-date. > + forage._dbInfo.db = dbInfo.db; > + forage._dbInfo.version = dbInfo.db; > + } > + } > + }); Should we just add in the catch and then merge this into master and address this use case later if needed then? I'd be okay with that... this patch is awesome. — Reply to this email directly or view it on GitHub <https://github.com/mozilla/localForage/pull/452/files#r40508172>.
-- *Thodoris Greasidis* Computer, Networking & Communications Engineer

Comment (sebweaver) [inline src/drivers/indexeddb.js]: My first idea was indeed to wait for other localForages to be ready, and by "ready" I mean "fulfilled" (sorry if my comment in the code wasn't clear enough about that). This goal of this implementation is to act as a kind of "transaction" actually. Thus, if several instance are created on the same database and one of them fails (like in @thgreasi's example) this implementation prevents, indeed, the next ones to be created. The idea behind that is to avoid any side-effect or inconsistency which might result from letting the process continuing. I mean, if multiples stores are involved in the same database, they are highly probably related somehow. And if an error is not explicitly handled at the initialization time, it has to be handled at a later time anyway.

That being said, I see your point. While my solution forces users to handle errors at initialization time, your solution to wait for localForages to be "not pending" (i.e. fulfilled or rejected) is indeed less restrictive and let them decide where and when to handle errors. So, on second thought, you might be right.

However, the way you suggest to achieve this goal can't work. Just adding an empty `catch()` on a promise doesn't change the status of the latter and returns it as is (i.e fulfilled or rejected). So the promise array will be globally fulfilled or rejected in the same way as before. On the other hand, a solution would be to return a resolved promise in both cases with:

```javascript
ready.push(forage.ready().catch(function() {
    return Promise.resolve();
});
```

Whereas this pattern could be confusing, if not commented, it gets the job done.

Now, the question which has to be decided is: what behavior do we want for users?

*Note: I hope my english is clear enough (here or in the code), it's not my native language...*

File: src/drivers/indexeddb.js (hunk 5 above)
Comment (sebweaver) [inline src/drivers/indexeddb.js]: Because `i` is already used above and jshint complains about that during the tests (with or without `var`). Since I didn't want to change the jshint config, I changed the name. :-)

**Comment by sebweaver** [comment]
> I have a few nitpicks about variable names and stuff

Sure, feel free to change them, or ask me to change them. English is not my native language, I won't be offended and be happy to improve things... ;-)

File: src/drivers/indexeddb.js (hunk 4 above)
Comment (thgreasi) [inline src/drivers/indexeddb.js]: Actually, [according to mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) `after a catch the chain is restored`, if I understood it right.

Comment (sebweaver) [inline src/drivers/indexeddb.js]: I admit it's a bit confusing, but "restoring the chain" doesn't mean changing the status of the original promise. And thankfully so, otherwise one can't be able to start different (and concurrent) chains on the same promise without massive side-effects between them. Promises are only, and exclusively, resolved or rejected by the executor given at their initialization.

Here is a [jsbin](https://jsbin.com/wujaki/edit?js,console) which illustrates this. It's based on the MDN example above and completed with our two arrays of promises. As one can see, even though `catch()` restores all chains in which it's been involved, only the second array `pAry2` is actually resolved.

Comment (thgreasi) [inline src/drivers/indexeddb.js]: You are right about `p2.catch().then(function(){ console.log('asdf'); })` not working
I never meant that a `p2.catch()` would alter the promise stored in variable `p2` (if not reassigned).

The fact is that `.catch()` will always return a new Promise so `p2 != p2.catch()`.

The part that confused me, was that it will not change the returned promise status, unless a function is passed as parameter (**any function!**).

In other words `Promise.reject('error').catch(function(){})` will always be resolved.
[Here is a jsbin with the above](https://jsbin.com/cofipapozi/edit?js,console)!?!?

*PS: That's a great discussion for a single PR!*

Comment (sebweaver) [inline src/drivers/indexeddb.js]: > I never meant that a p2.catch() would alter the promise stored in variable p2 (if not reassigned).

> The fact is that .catch() will always return a new Promise so p2 != p2.catch().

Fine! In fact, I just based my reasoning on your original suggestion using an empty catch vs. returning a new resolved promise. Sorry if inferred things that should not be. :blush: 

For the record, I've updated this [jsbin](https://jsbin.com/wujaki/edit?js,console) with the new case.

Now we are sure that en empty catch won't do the job unless any function is passed, the main question still remains: should the initialization process wait for others related instances to be fulfilled (current status of this PR) or just not pending?

As I said before, on second thought the second solution appears to be more flexible to users. If accepted, I can update the PR very soon.

Comment (thgreasi) [inline src/drivers/indexeddb.js]: I think that it should be @tofumatt 's decision. My call would be to go for the second case, where we check for `not pending`.
If that's how you will proceed, consider using `.catch(() => Promise.resolve())` for simplicity and consistency, as [found in other places in localForage source](https://github.com/mozilla/localForage/blob/086a344212124075cfba114caac1d842df78e5a0/src/localforage.js#L348).

Comment (tofumatt) [inline src/drivers/indexeddb.js]: (Sorry for my late reply). I'm in agreement here; the second solution is the more flexible option for developers and I think we should go with it.

Let me know when the PR is updated and I'll have one last look. :+1:

Comment (sebweaver) [inline src/drivers/indexeddb.js]: PR updated!
